h3/src/quic.rs

```diff
@@ -11,58 +11,100 @@ use bytes::Buf;
 //
 // - Should the `poll_` methods be `Pin<&mut Self>`?
 
+/// Trait that represent an error from the transport layer
+pub trait QuicError: std::error::Error + Send + Sync {
+    /// Check if the current error is a transport timeout
+    fn is_timeout(&self) -> bool;
+
+    /// Get the QUIC error code from connection close or stream stop
+    fn err_code(&self) -> Option<u64>;
+}
+
+impl<'a, E: QuicError + 'a> From<E> for Box<dyn QuicError + 'a> {
+    fn from(err: E) -> Box<dyn QuicError + 'a> {
+        Box::new(err)
+    }
+}
+
 /// Trait representing a QUIC connection.
 pub trait Connection<B: Buf> {
-    /// The stream type returned when opening a unidirectional send stream.
+    /// The type produced by `poll_accept_bidi()`
+    type BidiStream: SendStream<B> + RecvStream;
+    /// The type of the sending part of `BidiStream`
     type SendStream: SendStream<B>;
-    /// The stream type returned when accepting a unidirectional receive stream.
+    /// The type produced by `poll_accept_recv()`
     type RecvStream: RecvStream;
-    /// The stream type returned when accepting or opening a bidirectional
-    /// stream.
-    ///
-    /// This type doesn't *have* to implement `BidiStream`, but if the
-    /// implementation is able to provide a cheap split, it can opt-in.
-    type BidiStream: SendStream<B> + RecvStream;
-    /// The error type that can be returned when accepting or opening a stream.
-    type Error: Into<Box<dyn std::error::Error + Send + Sync>>;
-
-    // Accepting streams
+    /// A producer of outgoing Unidirectional and Bidirectional streams.
+    type OpenStreams: OpenStreams<B>;
+    /// Error type yeilded by this trait methods
+    type Error: Into<Box<dyn QuicError>>;
 
-    /// Poll the connection for any received bidirectional streams.
+    /// Accept an incoming unidirecional stream
     ///
     /// Returning `None` implies the connection is closing or closed.
-    fn poll_accept_bidi_stream(
+    fn poll_accept_recv(
```

Comment:
I feel like `_stream` here is automatically implied, because QUIC traits implementers and this crate developer actually know we are working with streams: this is the most basic knowledge about the protocol. Plus, the suffix is in every method name. And finally, `Stream` is in the name of the return type.